From: Jan Luebbe <jlu@pengutronix.de>
Date: Tue, 26 Jul 2022 14:47:10 +0200
Subject: [PATCH] lxatac: use custom PLL configuration
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We configure PLL3 to VCO = 624 MHz => P = 208, Q = 48, R = 104:
- 48 MHz for CAN
  20220726 mkl:
  > Bei Classical CAN basiert die Clock auf vielfachen von 8 MHz, das
  > spricht für 48 MHz (48/8=6, 78/8=9.75).
  >
  > Mit 48 MHz können bis auf die unübliche Bitrate von 666.666 kBit/s alle
  > standard Raten ohne Bitrate- und Sample Point Error erzeugt werden.
  >
  > Bei CAN-FD kommen wir mit 48 MHz bis 4 MBit/s ohne Fehler, bei 78 MHz
  > nur bis 2 MHz (aber mit Sample Point Error).
- 104 MHz for eMMC

We configure PLL4 to VCO = 750.0 MHz => P = 125, Q = 75, R = 62.5:
- 125 MHz for Ethernet

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>
---
 fdts/stm32mp157c-lxa-tac.dts | 52 ++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 50 insertions(+), 2 deletions(-)

diff --git a/fdts/stm32mp157c-lxa-tac.dts b/fdts/stm32mp157c-lxa-tac.dts
index 98286ac67aab..5988fded94da 100644
--- a/fdts/stm32mp157c-lxa-tac.dts
+++ b/fdts/stm32mp157c-lxa-tac.dts
@@ -80,7 +80,55 @@
 	};
 };
 
-/* Configure Eth PLL 125Mhz Clock output */
+/* VCO = 624 MHz => P = 208, Q = 48, R = 104 */
+&pll3 {
+	cfg = <1 51 2 12 5 PQR(1,1,1)>;
+	/delete-property/ frac;
+};
+
+/* VCO = 750.0 MHz => P = 125, Q = 75, R = 62.5 */
 &pll4 {
-	cfg = <3 124 5 11 11 PQR(1,1,1)>;
+	cfg = <3 124 5 9 11 PQR(1,1,1)>;
+	/delete-property/ frac;
+};
+
+&rcc {
+	/* change parent clocks */
+	st,pkcs = <
+		CLK_CKPER_HSE
+		CLK_FMC_ACLK
+		CLK_QSPI_ACLK
+		CLK_ETH_PLL4P
+		CLK_SDMMC12_PLL3R
+		CLK_DSI_DSIPLL
+		CLK_STGEN_HSE
+		CLK_USBPHY_HSE
+		CLK_SPI2S1_PLL3Q
+		CLK_SPI2S23_PLL3Q
+		CLK_SPI45_HSI
+		CLK_SPI6_HSI
+		CLK_I2C46_HSI
+		CLK_SDMMC3_DISABLED
+		CLK_USBO_USBPHY
+		CLK_ADC_CKPER
+		CLK_CEC_DISABLED
+		CLK_I2C12_HSI
+		CLK_I2C35_HSI
+		CLK_UART1_HSI
+		CLK_UART24_HSI
+		CLK_UART35_HSI
+		CLK_UART6_HSI
+		CLK_UART78_HSI
+		CLK_SPDIF_DISABLED
+		CLK_FDCAN_PLL3Q
+		CLK_SAI1_DISABLED
+		CLK_SAI2_DISABLED
+		CLK_SAI3_DISABLED
+		CLK_SAI4_DISABLED
+		CLK_RNG1_LSI
+		CLK_RNG2_LSI
+		CLK_LPTIM1_PCLK1
+		CLK_LPTIM23_PCLK3
+		CLK_LPTIM45_LSE
+	>;
 };
